DOC: AnnotationBbox keyword descriptions

rcomer opened the pull request:
## PR Summary

Closes #24390.

* Add a description of the `bboxprops` keyword for `AnnotationBbox` instantiation.  At https://github.com/matplotlib/matplotlib/issues/24390#issuecomment-1307083940 @oscargus noted that `arrowprops` and `fontsize` are also missing from the docstring, but possibly they are covered by "Other parameters are identical to Annotation"?  I note that the `arrowprops` description in [Annotation](https://matplotlib.org/stable/api/text_api.html#matplotlib.text.Annotation) is quite extensive, so am not sure of the wisdom of copying it across to related docstrings.
* Modify the AnnotationBbox demo to include some examples using the `bboxprops` keyword.  ~While I was in there I also changed all double quotes to single quotes, as there is currently a mixture and I think it should be consistent within the example.~

## PR Checklist

<!-- Please mark any checkboxes that do not apply to this PR as [N/A]. -->
**Tests and Styling**
- [N/A] Has pytest style unit tests (and `pytest` passes).
- [x] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).

**Documentation**
- [x] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [N/A] New plotting related features are documented with examples.

**Release Notes**
- [N/A] New features are marked with a `.. versionadded::` directive in the docstring and documented in `doc/users/next_whats_new/`
- [N/A] API changes are marked with a `.. versionchanged::` directive in the docstring and documented in `doc/api/next_api_changes/`
- [N/A] Release notes conform with instructions in  `next_whats_new/README.rst` or `next_api_changes/README.rst`

<!--
Thank you so much for your PR!  To help us review your contribution, please
consider the following points:

- A development guide is available at https://matplotlib.org/devdocs/devel/index.html.

- Help with git and github is available at
  https://matplotlib.org/devel/gitwash/development_workflow.html.

- Create a separate branch for your changes and open the PR from this branch. Please avoid working on `main`.

- The PR title should summarize the changes, for example "Raise ValueError on
  non-numeric input to set_xlim".  Avoid non-descriptive titles such as
  "Addresses issue #8576".

- The summary should provide at least 1-2 sentences describing the pull request
  in detail (Why is this change required?  What problem does it solve?) and
  link to any relevant issues.

- If you are contributing fixes to docstrings, please pay attention to
  http://matplotlib.org/devel/documenting_mpl.html#formatting.  In particular,
  note the difference between using single backquotes, double backquotes, and
  asterisks in the markup.

We understand that PRs can sometimes be overwhelming, especially as the
reviews start coming in.  Please let us know if the reviews are unclear or
the recommended next step seems overly demanding, if you would like help in
addressing a reviewer's comments, or if you have been waiting too long to hear
back on your PR.
-->

oscargus commented:
I agree that one should probably not copy the `arrowprops` part, but maybe something like
```
arrowprops: dict, optional
    Arrow properties, see `.Annotation` for description.

fontsize: float or str, optional
    Font size, see `~Text.set_fontsize`.
```
? 

As these properties are explicit in the constructor. (`fontsize` is documented as part of `**kwargs` in `Annotation` as "Additional kwargs are passed to Text" and not directly documented in `Text` either... One may wonder why it is explicitly mentioned in `AnnotationBbox`.

Turns out that `annotation_clip` is also missing...

Anyway, no strong opinions as such (I can always add those later), but I'll leave it for a while to see if someone has stronger opinions.

rcomer commented:
Now I look deeper, I think having "Other parameters are identical to Annotation" next to `**kwargs` here is also wrong, or at least misleading.  It implies that this `**kwargs` would include `Annotation`'s `**kwargs`, which are passed to `Text`, as you say.  But `AnnotationBbox`'s `**kwargs` are passed to `self._internal_update`.  So maybe the `**kwargs` description here should reference e.g. the [set](https://matplotlib.org/stable/api/offsetbox_api.html#matplotlib.offsetbox.AnnotationBbox.set) method?

timhoffm commented on examples/text_labels_and_annotations/demo_annotation_box.py:
```diff
@@ -22,27 +22,28 @@
 
 # Define a 1st position to annotate (display it with a marker)
 xy = (0.5, 0.7)
-ax.plot(xy[0], xy[1], ".r")
```
There's a lot of quote changes. Is that intentional? Please note that we currently do not have a convention on single vs. double quotes. Both are acceptable, but we don't want to switch back and forth in existing code.

rcomer replied:
Fair enough.  It was just to make sure there was one consistent type within this script, and I thought I had seen more single than double within Matplotlib.  Happy to change it back if that is not desired.

timhoffm replied:
Yes please. If we go for consistent string styling this will need a separate discussion and a decision which way to go (e.g. black enforces double quotes). Until now, let's keep existing quotes as they are.

rcomer replied:
Done.

rcomer commented:
I think I have addressed all the comments, but...

What text does this `fontsize` keyword apply to?  If I put e.g. `fontsize="xx-large"` in the call to `AnnotationBbox` in my example at #24390, it doesn't seem to make a difference.  That makes sense, as you would set the fontsize for the text in the box using the `TextArea` object instead.  Is there some other text associated with the `AnnotationBbox`?

timhoffm commented:
> What text does this `fontsize` keyword apply to?

Looking into the code, the fontsize parameter is stored in a `FontProperties`object at `AnnotationBbox.prop` and its only use is to define the `mutation_scale` parameter of the [FancyBboxPatch](https://matplotlib.org/devdocs/api/_as_gen/matplotlib.patches.FancyBboxPatch.html), i.e. affecting attributes of the box style like padding and rounding of corners (if defined).

AFAICS we should rename `fontsize` to `mutation_scale` and deprecate `prop`. But ping @anntzer, who wrote this code in 2009.

anntzer commented:
I didn't really look at what piece of code you're talking about (can you point out the line?) but I was certainly not involved with matplotlib development yet in 2009...

rcomer commented:
I think `mutation_scale` can be passed as part of `bboxprops`, so we could just deprecate `fontsize` and direct the user to use `bboxprops=dict(mutation_scale=x)` instead?  It would seem odd to me to have `mutation_scale` accessible at the top level when the other `FancyBboxPatch` parameters (like `boxstyle`) are passed within the dictionary.

https://github.com/matplotlib/matplotlib/blob/1fa74675ac38f768090525b6acad1fecd7db61f6/lib/matplotlib/offsetbox.py#L1296-L1305

timhoffm commented:
> I didn't really look at what piece of code you're talking about (can you point out the line?) but I was certainly not involved with matplotlib development yet in 2009...

Sorry, mixed up users. - Just read the short version "09.08.09 Lee", but it was @leejjoon. @leejjoon do you rember the motivation in `AnnotationBbox` to pass `fontsize` -> `self.prop` -> `FancyBboxPatch(..., mutation_scale=...)`

https://github.com/matplotlib/matplotlib/blob/1fa74675ac38f768090525b6acad1fecd7db61f6/lib/matplotlib/offsetbox.py#L1235

https://github.com/matplotlib/matplotlib/blob/1fa74675ac38f768090525b6acad1fecd7db61f6/lib/matplotlib/offsetbox.py#L1348-L1358

https://github.com/matplotlib/matplotlib/blob/1fa74675ac38f768090525b6acad1fecd7db61f6/lib/matplotlib/offsetbox.py#L1296-L1302

it indeed seems that `mutation_scale` can be passed already via `bboxprops` and the `fontsize` parameter is not needed at all.

rcomer commented:
I guess using `fontsize` gives the user the convenience of passing a string, rather than points.  Also it currently defaults to `rcParams["legend.fontsize"]`.  If we left it to the `FancyBBoxPatch` default of 1, it would presumably look quite cramped.

So perhaps using `fontsize` is the most practical way to get a nice looking pad, etc., even though we are not dealing with any actual fonts.

leejjoon commented:
My vague recollection is that I wanted to copy the behavior from annotation (and text) that the mutation scale is set to their fontsize. As @rcomer commented in his post, I think this was a sensible thing to do. The parameter name of `fontsize` could be misleading and may need to be replaced. It would be good if we can keep the current behavior, though.

rcomer commented:
@timhoffm I have updated the `fontsize` description to explain how it is used.  Do you think this is enough?  Once we understand the reasoning, the name does not seem _wrong_ to me.  Perhaps a little unintuitive, but I wonder if that matters enough to go through the breaking change procedure.

timhoffm commented on lib/matplotlib/offsetbox.py:
```diff
@@ -1270,8 +1275,24 @@ def __init__(self, offsetbox, xy,
+        fontsize: float or str, optional
+            Translated to points and passed as *mutation_scale* into
+            `.FancyBboxPatch` to scale attributes of the box style (e.g. pad
+            or rounding_size).  Passed here as *fontsize* for the convenience
+            of using the named sizes.  If not given, :rc:`legend.fontsize` is
```
```suggestion
            or rounding_size).  If not given, :rc:`legend.fontsize` is
```
I'd leave this out. I think it's not really helpful for understanding (or even technically necessary to provide the convenience).

If you want to comment on the name, it should be more in the direction 

> the name is chosen in analogy to [...*] where fontsize defines the mutation scale as well.

\* To be checked what exactly: Text/Annotation/Legend/FancyBboxPatch.

rcomer replied:
Looks like it's in `Text`
https://github.com/matplotlib/matplotlib/blob/57710dba1ac30be50e2c6bdaa691036a3354ca4a/lib/matplotlib/text.py#L582

timhoffm commented on lib/matplotlib/offsetbox.py:
```diff
@@ -1270,8 +1275,24 @@ def __init__(self, offsetbox, xy,
+
         **kwargs
-            Other parameters are identical to `.Annotation`.
+            Other `AnnotationBbox` properties.  See `~.AnnotationBbox.set` for
```
```suggestion
            Other `AnnotationBbox` properties.  See `.AnnotationBbox.set` for
```
Minor, but I think it's slightly better to display the class here as well.

timhoffm reviewed (commented):
> @timhoffm I have updated the `fontsize` description to explain how it is used.  Do you think this is enough?  Once we understand the reasoning, the name does not seem _wrong_ to me.  Perhaps a little unintuitive, but I wonder if that matters enough to go through the breaking change procedure.

I still think the name is misleading because it does not affect any font. It only happens to be that mutation_scale is defined in such a way that setting mutation_scale to the used fontsize gives a reasonable scaling (e.g. for pad). That *could* also be done with an explicit name *mutation_scale*, including supporting string values.

That said, I would not name it like this again, but I'm not dogmatic about changing the name. When we have at least documented what the parameter is about that's already a big improvement. But if we keep the name `AnnotationBbox.set_fontsize` needs a similar description.

timhoffm commented:
Thanks @rcomer! I appreciate this work. It's often more cumbersome than initially expected, but it really makes our docs much better.
